--- CMakeLists.txt.orig 2016-02-22 11:29:56.212975415 +0000
+++ CMakeLists.txt      2016-02-22 11:33:07.747090366 +0000
@@ -382,7 +382,7 @@
     INSTALL(DIRECTORY include/ DESTINATION include FILES_MATCHING PATTERN "*.h")
 ENDIF ()

-INSTALL(PROGRAMS share/h2o/annotate-backtrace-symbols share/h2o/fetch-ocsp-response share/h2o/kill-on-close share/h2o/setuidgid share/h2o/start_server DESTINATION share/h2o)
+INSTALL(PROGRAMS share/h2o/annotate-backtrace-symbols share/h2o/fastcgi-cgi share/h2o/fetch-ocsp-response share/h2o/kill-on-close share/h2o/setuidgid share/h2o/start_server DESTINATION share/h2o)
 INSTALL(DIRECTORY doc/ DESTINATION share/doc/h2o PATTERN "Makefile" EXCLUDE PATTERN "README.md" EXCLUDE)

 # tests
